util-split.R

```
@@ -99,26 +151,35 @@ split.data.time.based = function(project.data, time.period = "3 months", bins =
         ## remove sliding windows
         sliding.window = FALSE
     }
+
+    ## indicates if time-based splitting is performed using bins
+    split.time.based.with.bins = FALSE
+
     ## if bins are NOT given explicitly
     if (is.null(bins)) {
         ## get bins based on split.basis
-        bins = split.get.bins.time.based(data[[split.basis]][["date"]], time.period, number.windows)$bins
+        bins = split.get.bins.time.based(data[[split.basis]][["date"]], splitting.length, number.windows)$bins
         bins.labels = head(bins, -1)
-        split.by.bins = FALSE
         ## logging
         logging::loginfo("Splitting data '%s' into time ranges of %s based on '%s' data.",
-                         project.data$get.class.name(), time.period, split.basis)
+                         project.data$get.class.name(), splitting.length, split.basis)
     }
-    ## when bins are given explicitly
+    ## when bins are given explicitly, get bins based on parameter
     else {
-        ## remove sliding windows
-        sliding.window = FALSE
-        ## get bins based on parameter
-        split.basis = NULL
-        bins = get.date.from.string(bins)
-        bins = get.date.string(bins)
+        if (split.by.time) {
+            split.time.based.with.bins = TRUE
+            split.basis = NULL
+            bins = get.date.from.string(bins)
+            bins = get.date.string(bins)
+            ## remove sliding windows
+            sliding.window = FALSE
+        } else {
+            ## sliding windows do not need to be removed here, as sliding windows and bins
+            ## are not contradicting in activity-based splitting
+            bins.vector = bins[["vector"]]
+            bins = bins[["bins"]]
+        }
```

Please allow me to be forgetful. We already have discussed about these lines of code some weeks ago, but when I look at them again, I again don't understand what's going on here:

This "else" case handles the case that we want to split data activity-based by given bins, right?
How does this work? How can we split activity-based by bins? I could imagine that we can count the activities between two bins and use this activity amount for activity-based splitting. but what do we do when we have more than two bins given? This confuses me. And then the comment also states that we can have sliding windows in addition - how could this work if I cannot even imagine it without sliding windows. Maybe I am too tired now to be able to understand this little piece of code, but I am really confused now. 

Can you enlighten me in this matter @MaLoefUDS ?

---

Sure I think I can explain it.

Your first assumption is correct. When doing activity based splitting, we split all our data by a given `split.basis`, like `issue`, into sections with n unique entries. Now we expand the vectors to include all entries with duplicate ids as well. This means that the bins are now not all the same size. Those bins are then passed into the referenced `split.data.by.time.or.bins` method. This method then just splits the data by the given bins (not only the data of the given `split.basis` but *all* data. 

The role of `sliding.window` in this function when used for activity based splitting is a bit odd. The parameter does in this case **not** decide the splitting algorithm or similar, so the splitting stays the same whether this parameter is true or false however this method can be called with bins representing regular bins or representing sliding window bins and depending on this the splitting info in the end needs to be adjusted to fit. The comment actually refers to a change you requested. As in the then case of the if we set `sliding.window` to false because we cannot have given bins in time-based splitting and use sliding windows at the same time. In the case of activity based splitting these two do not contradict, in fact the parameter is not allowed to be overwritten there as we need it in the end for the splitting info. 

Hope that explains it :)

---

Ah, I get it. Thanks for the detailed explanations. Then everything is fine. I just did not question where the bins come from in this case. But if they are already given as function parameter correctly, then everything makes sense.
Can we maybe add an additional comment at the beginning of the else block? Something like
"the activity-based bins are already passed to this function in the correct format"
But as we do not make sure that the parameter is in the correct format, maybe "are" is wrong - this is just an assumption which may not hold when someone would use this function incorrectly. So maybe we should go with "should be" instead?

Should we check whether format is correct and print a warning if not? What is your opinion on that?

---

The method, which was the base of the new `split.data.by.time.or.bins` method, did no checks on the format of the bins either, even though the multiple notations of bins used here was not introduced by me. Other parameters are not checked consequently for their format either, and lastely, the method is only called by the two new wrapper methods, so if we want to introduce the format checks, I think we should do it there :).

---

Ok, thanks for your reply. I am of two minds. Some time ago, we started to do input validation in some of the methods of the Conf and Data classes. I am not sure if we also should do it for other functions, such as the splitting functions, or if we should just let it fail if the format is wrong.

@hechtlC What is your opinion on that?